pvlib/ivtools.py

```diff
@@ -0,0 +1,262 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Mar 29 10:34:10 2019
+
+@author: cwhanse
+"""
+
+import numpy as np
+
+
+def fit_cec_with_sam(celltype, v_mp, i_mp, v_oc, i_sc, alpha_sc, beta_voc,
+                     gamma_pmp, cells_in_series, temp_ref=25):
+    '''
+    Estimates parameters for the CEC single diode model using the SAM SDK.
+
+    Parameters
+    ----------
+    celltype : str
+        Value is one of 'monoSi', 'multiSi', 'polySi', 'cis', 'cigs', 'cdte',
+        'amorphous'
+    v_mp : float
+        Voltage at maximum power point at standard test condition (STC)
+    i_mp : float
+        Current at maximum power point at STC
+    Voc : float
+        Open circuit voltage at STC
+    i_sc : float
+        Short circuit current at STC
+    alpha_sc : float
+        Temperature coefficient of short circuit current at STC, A/C
+    beta_voc : float
+        Temperature coefficient of open circuit voltage at STC, V/C
+    gamma_pmp : float
+        Temperature coefficient of power at maximum point point at STC, %/C
+    cells_in_series : int
+        Number of cells in series
+    temp_ref : float, default 25
+        Reference temperature condition
+
+    Returns
+    -------
+    a_ref : float
+        The product of the usual diode ideality factor (n, unitless),
+        number of cells in series (Ns), and cell thermal voltage at reference
+        conditions, in units of V.
+
+    I_L_ref : float
+        The light-generated current (or photocurrent) at reference conditions,
+        in amperes.
+
+    I_o_ref : float
+        The dark or diode reverse saturation current at reference conditions,
+        in amperes.
+
+    R_sh_ref : float
+        The shunt resistance at reference conditions, in ohms.
+
+    R_s : float
+        The series resistance at reference conditions, in ohms.
+
+    Adjust : float
+        The adjustment to the temperature coefficient for short circuit
+        current, in percent
+
+    Raises:
+        ImportError if NREL-PySAM is not installed
+    '''
+
+    try:
+        from PySAM import PySSC
+    except ImportError as e:
+        raise(e)
```

Comment:
I don't think the try-except is necessary.

Reply:
Perhaps not. I don't know how the code performs if PySAM isn't found. I think explicitly raising `ImportError` could help. One challenge, there is a python package `pysam` and then there's `NREL-PySAM`, it's the latter that is needed. Do you think that raising a specific message (rather than the `ImportError`) would help here?

Reply:
It looks like a `ModuleNotFoundError` exception (not `ImportError`) is raised if the package is not installed:
```terminal
$ python -c "from PySAM import PySSC"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'PySAM'
```
Because of the naming similarity issue, it might be worth "raising further" from within the original exception with a message that points to NREL's package. This way you don't clobber the original exception's information, which might be useful for debugging the problem. Perhaps something like:
```python
try:
    from PySAM import PySSC
except Exception as e:
        raise ImportError("Make sure you have NREL's PySAM package installed correctly. It is found at https://pypi.org/project/NREL-PySAM/.") from e
```
See https://stackoverflow.com/questions/6062576/adding-information-to-an-exception

Reply:
I like @markcampanelli's suggestion with a minor change: `ModuleNotFoundError` is a subclass of `ImportError`, so `except ImportError` will work for any kind of import problem.

Reply:
Sorry I missed the subclass relationship. I'm ever the skeptic, so I'm less comfortable limiting the scope of what could have gone wrong with the import. As long as the other exceptions don't get swallowed I'm fine with limiting the except clause to only `ImportError`s.

Reply:
Done. Thanks @markcampanelli